File: 2021/11/11.fs
module _11

// PARSING

let lines = System.IO.File.ReadLines("11/test.txt") |> Seq.toArray

let grid = lines |> Array.map (fun line -> line |> Seq.map (fun c -> int c - 48) |> Seq.toArray)

// PART 1

let editGrid pred f = Array.iteri (fun y row -> row |> Array.iteri (fun x n -> if pred x y n then do row[x] <- f n))

let getAdjacent x y = Array.allPairs [|(max x 0)..(min x 9)|] [|(max y 0)..(min y 9)|]

let getFlashes = Array.mapi (fun y row -> row |> Array.mapi (fun x n -> if n > 9 then getAdjacent x y else [||]) |> Array.concat ) >> Array.concat >> Array.distinct

let rec subStep flashes grid = 
    let newFlashes = Array.except flashes <| getFlashes grid
    if newFlashes.Length = 0 
    then
        do editGrid (fun _ _ n -> n < 0) (fun _ -> 0) grid
        0
    else
        do editGrid (fun x y _ -> Array.contains (x, y) newFlashes) (fun n -> n + 1) grid
        newFlashes.Length + subStep (Array.append flashes newFlashes) grid

let step grid = 
    editGrid (fun _ _ n -> n > 9) (fun n -> 0) grid //reset
    editGrid (fun _ _ _ -> true) (fun n -> n + 1) grid //increment
    subStep [||] grid

let octopus n grid = [|0..n|] |> Array.map (fun _ -> step grid)

let test = octopus 10 grid

let part1 = sprintf "%d" <| 0

// PART 2

let part2 = sprintf "%d" <| 0

File: 2021/11/11.py
from functools import reduce

def line_to_array(line) :
    nums = []
    for c in line:
        nums.append(int(c))
    return nums

f = open("11.txt", "r")
lines = f.read().split("\n")
array = list(map(line_to_array, lines))

def adjacent(y, x):
    return [
        (y-1, x-1),
        (y-1, x),
        (y-1, x+1),
        (y, x-1),
        (y, x+1),
        (y+1, x-1),
        (y+1, x),
        (y+1, x+1),
        ]

def flash(y, x, array, skip):
    if y < 0 or y >= len(array) or x < 0 or x >= len(array[0]):
        return 0
    if not skip:
        array[y][x] += 1
    flashes = 0
    if array[y][x] == 10:
        array[y][x] += 1
        flashes += 1
        for adj in adjacent(y,x):
            flashes += flash(adj[0], adj[1], array, skip)
    return flashes

def increment(array):
    for y in range(len(array)):
        for x in range(len(array[y])):
            array[y][x] += 1

def reset(array):
    resets = 0
    for y in range(len(array)):
        for x in range(len(array[y])):
            if array[y][x] > 9:
                array[y][x] = 0
                resets += 1
    return resets

                
def step(array):
    flashes = 0
    skip = False
    last_flashes = 1
    while last_flashes > 0:
        last_flashes = 0
        for y in range(len(array)):
            for x in range(len(array[y])):
                last_flashes += flash(x,y,array, skip)
        flashes += last_flashes
        skip = True
    resets = reset(array)
    return resets
    # return flashes
                

# steps = 100
# flashes = 0
# for i in range(steps):
#     flashes += step(array)
# print(flashes)

steps = 1
while (step(array) != 100):
    steps += 1

print(steps)
